### PR TITLE
Not allocating `Writeable`s for regex evaluation

### PR DESCRIPTION
**Pull request by robertbastian**
Depends on #2994

### CODE REVIEW

File: components/list/src/lazy_automaton.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use regex_automata::dfa::sparse::DFA;
+use regex_automata::dfa::Automaton;
+use regex_automata::util::id::StateID;
+use writeable::Writeable;
+
+pub trait LazyAutomaton: Automaton {
+    // Like Automaton::find_earliest_fwd, but doesn't require a materialized string.
```
Comment (sffc) [inline components/list/src/lazy_automaton.rs:11]: Question: `find_earliest_fwd` says that it searches for the first match, whereas this function appears to require that the match appear at the beginning of the string; is that correct?

Comment (robertbastian) [reply components/list/src/lazy_automaton.rs:11]: The DFAs we construct are all "anchored", so only match at the beginning.

Comment (robertbastian) [reply components/list/src/lazy_automaton.rs:11]: See the test

Comment (sffc) [reply components/list/src/lazy_automaton.rs:11]: OK I think I get it now.

**Review by sffc — APPROVED** [review]

**Pull request merged**